tools/kafka-versions-tools.sh

```diff
@@ -38,6 +38,70 @@ function get_kafka_versions {
 
 }
 
+function get_kafka_source_urls {
+
+    eval kafka_source_urls="$(yq read $VERSIONS_FILE '*.url' -j | tr '[],' '() ')"
+
+}
+
+function get_kafka_source_files {
+
+    eval kafka_source_files="$(yq read $VERSIONS_FILE '*.file' -j | tr '[],' '() ')"
+
+}
+
+function get_url_file_map {
+    # This function loops over the yaml objects for each version, rather the just using a single line 
+    # wild card, because the url and file entries are not guaranteed to always be present and 
+    # therefore the length of their wild card arrays may not match the lengths of versions array.
+
+    declare -Ag version_binary_urls
+    declare -Ag version_binary_files
+
+    finished=0
+    counter=0
+    while [ $finished -lt 1 ] 
+    do
+        version="$(yq read $VERSIONS_FILE [${counter}].version)"
+
+        if [ "$version" = "null" ]
+        then
+            finished=1
+        else
+            source_count=0
+
+            url="$(yq read $VERSIONS_FILE [${counter}].url)"
+            file="$(yq read $VERSIONS_FILE [${counter}].file)"
+            
+            if [ "$url" != null ]
+            then
+                version_binary_urls["${version}"]="$url"
+                source_count=$((source_count+1))
+            fi
+
+            if [ "$file" != null ]
+            then
+                version_binary_files["${version}"]="$file"
+                source_count=$((source_count+1))
+            fi
+
+            if [ $source_count == 0 ]
+            then 
+                echo -e "No binary source specified for Kafka $version. Either 'url' or 'file' fields must be specified in the kafka-versions.yaml file"
+                exit 1
+            fi
```

Comment:
What's wrong with `if/else if/else`

Reply:
I wanted to populate both arrays in this general tools method and then leave it up to the caller which one you want to take. `if/else if` would give one precedent over the other and loose the info.

Reply:
You're providing these functions to help the rest of the build, and you've documented in the yaml itself that there's an order of precedence behaviour. So I think it's fine that the tools you're providing enforce that.

Reply:
A fair point. I'll change it.